### PR TITLE
[1LP][RFR] Cloud tenant collection

### PR DESCRIPTION
**Pull request by vatsalparekh**
Purpose or Intent
=================
Collection implementation for Cloud Tenants + its tests
{{pytest: cfme/tests/cloud/test_tenant.py -vvvv}}

### CODE REVIEW

File: cfme/cloud/tenant.py [hunk 1]
```diff
@@ -55,7 +54,8 @@ class TenantEntities(View):
     title = Text('//div[@id="main-content"]//h1')
     table = Table('//div[@id="list_grid"]//table')
     search = View.nested(Search)
-    # element attributes changed from id to class in upstream-fine+, capture both with locator
+    # element attributes changed from id to class in upstream-fine+, capture
+    # both with locator
```
Comment (mshriver) [inline cfme/cloud/tenant.py]: Why is this split on 2 lines? Its only 94 chars.  I'd rather see the word 'locator' removed than split this if I'm counting wrong and it is over 100.

File: cfme/cloud/tenant.py [hunk 2]
```diff
@@ -92,7 +93,8 @@ def in_tenants(self):
         return (
             self.logged_in_as_current_user and
             self.navigation.currently_selected == ['Compute', 'Clouds', 'Tenants'] and
-            # TODO: Needs to be converted once there's a Widgetastic alternative
+            # TODO: Needs to be converted once there's a Widgetastic
```
Comment (mshriver) [inline cfme/cloud/tenant.py]: Again don't split comment lines unless they're over 100 chars and lint is complaining.

**Comment by mshriver** [comment]
Looks like an IDE misconfig cut a bunch of statements onto 2 lines on an 80 char limit instead of 100. Spoke with @vatsalparekh and he's going to be updating branch to resolve these back to 100 char limit before review.

File: cfme/cloud/tenant.py [hunk 3]
```diff
@@ -194,29 +193,35 @@ def is_displayed(self):
...
-    _param_name = 'Tenant'
+class TenantCollection(Navigatable):
+    """Collection object for the :py:class:`cfme.cloud.tenant.Tenant`."""
 
-    def __init__(self, name, provider, appliance=None):
-        """Base class for a Tenant"""
-        self.name = name
-        self.provider = provider
-        Navigatable.__init__(self, appliance=appliance)
+    def create(self, name, provider, applience=None, cancel=False, wait=False):
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
Misspelling is here `applience`

File: cfme/cloud/tenant.py [hunk 4]
```diff
@@ -194,29 +193,35 @@ def is_displayed(self):
...
+        page.form.fill({
+            'cloud_provider': provider.name,
+            'name': name
         })
         if cancel:
-            view.form.cancel_button.click()
+            page.form.cancel_button.click()
         else:
-            view.form.save_button.click()
+            page.form.save_button.click()
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
What will happen if the form is not filled and `cancel` is `False`?

File: cfme/cloud/tenant.py [hunk 5]
```diff
@@ -194,29 +193,35 @@ def is_displayed(self):
...
 
         all_view = self.create_view(TenantAllView)
         wait_for(lambda: all_view.is_displayed, fail_condition=False, num_sec=120, delay=3,
                  fail_func=lambda: all_view.flush_widget_cache(), handle_exception=True)
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
```python
wait_for(
    lambda: all_view.is_displayed,
    num_sec=120,
    delay=3,
    fail_func=all_view.flush_widget_cache,
    handle_exception=True
)
```
```

File: cfme/cloud/tenant.py [hunk 6]
```diff
@@ -225,17 +230,67 @@ def create(self, cancel=False, wait=False):
...
+
+    def delete(self, *tenants):
+        """Delete one or more  Tenants from the list of the Tenants
+
+        Args:
+            list of the `cfme.cloud.tenant.Tenant` objects
+        """
+        if self.appliance.version < '5.7':
+            raise OptionNotAvailable(
+                'Cannot delete cloud tenants in CFME < 5.7')
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
This check is not needed

File: cfme/cloud/tenant.py [hunk 7]
```diff
@@ -263,43 +320,26 @@ def update(self, updates, wait=True):
...
-            view.toolbar.view_selector.select('List View')
-            try:
-                row = view.paginator.find_row_on_pages(view.entities.table, name=self.name)
-                row[0].check()
-                view.toolbar.configuration.item_select('Delete Cloud Tenants',
-                                                       handle_alert=not cancel)
-            except NoSuchElementException:
-                raise TenantNotFound('Tenant {} not found'.format(self.name))
+            raise OptionNotAvailable(
+                'Cannot delete cloud tenants in CFME < 5.7')
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
Not needed

File: cfme/cloud/tenant.py [hunk 8]
```diff
@@ -263,43 +320,26 @@ def update(self, updates, wait=True):
...
 
-        if cancel:
-            return self.exists
-        else:
-            # Flash message is the same whether deleted from details or by selection
-            result = view.entities.flash.assert_success_message(
+        try:
+            view = navigate_to(self, 'Details')
+        except Exception as ex:
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
Please specify the exception.

File: cfme/cloud/tenant.py [hunk 9]
```diff
@@ -310,14 +350,15 @@ def exists(self):
...
+@navigator.register(Server, 'All')
 class TenantAll(CFMENavigateStep):
     VIEW = TenantAllView
     prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
 
     def step(self, *args, **kwargs):
         """Go to the All page"""
-        self.prerequisite_view.navigation.select('Compute', 'Clouds', 'Tenants')
+        self.prerequisite_view.navigation.select(
+            'Compute', 'Clouds', 'Tenants')
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
Don't break the line

File: cfme/cloud/tenant.py (hunk 4 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Changed as per your suggestion, but then what will happen now in the case when the form is filled and cancel is given True?

File: cfme/cloud/tenant.py [hunk 10]
```diff
@@ -1,3 +1,5 @@
+from __future__ import print_function
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
Why is this here?

File: cfme/cloud/tenant.py [hunk 11]
```diff
@@ -194,29 +195,35 @@ def is_displayed(self):
...
-    _param_name = 'Tenant'
+class TenantCollection(Navigatable):
+    """Collection object for the :py:class:`cfme.cloud.tenant.Tenant`."""
 
-    def __init__(self, name, provider, appliance=None):
-        """Base class for a Tenant"""
-        self.name = name
-        self.provider = provider
-        Navigatable.__init__(self, appliance=appliance)
+    def create(self, name, provider, appliance=None, cancel=False, wait=False):
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
`cancel` should be removed from the arguments.

File: cfme/cloud/tenant.py [hunk 12]
```diff
@@ -194,29 +195,35 @@ def is_displayed(self):
...
-            view.form.save_button.click()
+            page.form.cancel_button.click()
 
         all_view = self.create_view(TenantAllView)
-        wait_for(lambda: all_view.is_displayed, fail_condition=False, num_sec=120, delay=3,
-                 fail_func=lambda: all_view.flush_widget_cache(), handle_exception=True)
+        wait_for(lambda: all_view.is_displayed, num_sec=120, delay=3,
+                 fail_func=all_view.flush_widget_cache, handle_exception=True)
+
         if cancel:
```
Comment (quarckster) [inline cfme/cloud/tenant.py]:
```python
if not changed
```

File: cfme/cloud/tenant.py (hunk 10 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: by mistake, corrected

File: cfme/cloud/tenant.py (hunk 11 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: corrected

File: cfme/cloud/tenant.py [hunk 13]
```diff
@@ -194,48 +193,101 @@ def is_displayed(self):
...
-        self.name = name
-        self.provider = provider
-        Navigatable.__init__(self, appliance=appliance)
+    def create(self, name, provider, appliance=None, wait=False):
+        '''Add a cloud Tenant from the UI and return the Tenant object
+
+        Args:
+            name(str): name of the tenant to add
+            provider(obj): provider object on which you want to add the tenant
+            cancel(bool): True if you want to click cancel after filling the tenant form
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: remove it please

File: cfme/cloud/tenant.py [hunk 14]
```diff
@@ -194,48 +193,101 @@ def is_displayed(self):
...
+class TenantCollection(Navigatable):
+    """Collection object for the :py:class:`cfme.cloud.tenant.Tenant`."""
 
-    def __init__(self, name, provider, appliance=None):
-        """Base class for a Tenant"""
-        self.name = name
-        self.provider = provider
-        Navigatable.__init__(self, appliance=appliance)
+    def create(self, name, provider, appliance=None, wait=False):
+        '''Add a cloud Tenant from the UI and return the Tenant object
```
Comment (psav) [inline cfme/cloud/tenant.py]: Can you use ```"""``` and not ```'''``` for docstring comments please as per the guidelines

File: cfme/cloud/tenant.py [hunk 15]
```diff
@@ -254,6 +306,8 @@ def wait_for_appear(self, timeout=600):
                         message='Wait for cloud tenant to appear', fail_func=self.browser.refresh)
 
     def update(self, updates, wait=True):
+        if not self.exists:
```
Comment (psav) [inline cfme/cloud/tenant.py]: **REQUIRED**: No, we shouldn't put this here, if it goes anywhere, it goes at the end of the create.

File: cfme/cloud/tenant.py [hunk 16]
```diff
@@ -310,7 +344,7 @@ def exists(self):
             return False
 
 
-@navigator.register(Tenant, 'All')
+@navigator.register(Server, 'All')
```
Comment (psav) [inline cfme/cloud/tenant.py]: **REQUIRED**: ```TenantAll``` should be registered against ```TenantCollection```

File: cfme/tests/cloud/test_tenant.py [hunk 17]
```diff
@@ -13,13 +13,14 @@
...
-    tenant = Tenant(name=fauxfactory.gen_alphanumeric(8), provider=provider)
+    collection = TenantCollection()
+    tenant = collection.create(name=fauxfactory.gen_alphanumeric(8), provider=provider)
 
     yield tenant
 
     try:
         if tenant.exists:
-            tenant.delete()
+            collection.delete(tenant)
```
Comment (psav) [inline cfme/tests/cloud/test_tenant.py]: Why move this to the collection?

File: cfme/cloud/tenant.py (hunk 14 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Updated. But can you help me with which guidelines you refer to?

File: cfme/cloud/tenant.py (hunk 13 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Removed

File: cfme/cloud/tenant.py (hunk 15 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Removed it, and in `create` I made the `wait=True` by default, because tenants take some time to appear.

File: cfme/cloud/tenant.py (hunk 16 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Changed

File: cfme/tests/cloud/test_tenant.py (hunk 17 above)
Comment (vatsalparekh) [inline cfme/tests/cloud/test_tenant.py]: No specific reason, will it revert back

**Comment by dajoRH** [comment]
Lint report for commit 3f5de60403bbc5f542ea551ac26280699c3cfaef:

`cfme/cloud/tenant.py`:
- :red_circle: Line 15:1: **F401** *'cfme.base.Server' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2e824601d4a42644c8441ed026efde73445670f1:

`cfme/cloud/tenant.py`:
- :bangbang: Line 235:25: **E127** *continuation line over-indented for visual indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 72df06342e6e9926cf320699ff40237a805dff8e:

`cfme/cloud/tenant.py`:
- :red_circle: Line 23:1: **F401** *'pdb' imported but unused*
- :bangbang: Line 234:25: **E127** *continuation line over-indented for visual indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit c4fd7411141106443344b0ab213f6d49e1ff0e27:

`cfme/cloud/tenant.py`:
- :red_circle: Line 23:1: **F401** *'pdb' imported but unused*
- :bangbang: Line 234:25: **E127** *continuation line over-indented for visual indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2a48c42adfe4f27e76059d592ada743ec4f3fbd9:

`cfme/cloud/tenant.py`:
- :bangbang: Line 235:25: **E127** *continuation line over-indented for visual indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/cloud/test_tenant.py (hunk 17 above)
Comment (vatsalparekh) [inline cfme/tests/cloud/test_tenant.py]: Maybe should keep it to collection, it does same thing, but to check the PR

File: cfme/cloud/tenant.py (hunk 15 above)
Comment (psav) [inline cfme/cloud/tenant.py]: OK, Not seeing it appear here yet though

File: cfme/tests/cloud/test_tenant.py (hunk 17 above)
Comment (psav) [inline cfme/tests/cloud/test_tenant.py]: No, let's revert please

File: cfme/cloud/tenant.py (hunk 15 above)
Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Sorry for the half fix, I've updated it now.

File: cfme/tests/cloud/test_tenant.py (hunk 17 above)
Comment (vatsalparekh) [inline cfme/tests/cloud/test_tenant.py]: Okay, for Collections, I should add a separate test then

File: cfme/cloud/tenant.py [hunk 18]
```diff
@@ -194,48 +192,93 @@ def is_displayed(self):
...
+                tenant.provider.refresh_provider_relationships()
                 all_view.flush_widget_cache()
                 self.browser.refresh()
 
-            wait_for(lambda: self.exists, timeout=600, message='Wait for cloud tenant to appear',
+            wait_for(lambda: tenant.exists, timeout=600,
+                     message='Wait for cloud tenant to appear',
                      delay=10, fail_func=refresh)
 
+        return tenant
```
Comment (mfalesni) [inline cfme/cloud/tenant.py]: There should be an `.instantiate` method that creates the tenant object.

File: cfme/cloud/tenant.py [hunk 19]
```diff
@@ -194,48 +192,93 @@ def is_displayed(self):
...
+
+        # TODO: Assert deletion flash message for selected tenants
+        # it is not shown in current UI, so not asserting
+
+
+class Tenant(Navigatable):
+    '''Tenant Class'''
+    _param_name = 'Tenant'
+
+    def __init__(self, name, provider, appliance=None):
```
Comment (mfalesni) [inline cfme/cloud/tenant.py]: This should accept the parent collection instead of an appliance, getting the appliance from the collection.

File: cfme/cloud/tenant.py [hunk 20]
```diff
@@ -194,48 +192,97 @@ def is_displayed(self):
...
                 msg = 'Add of Cloud Tenant was cancelled by the user'
             else:
                 msg = 'Add of new Cloud Tenant was cancelled by the user'
             all_view.entities.flash.assert_success_message(msg)
         else:
             all_view.entities.flash.assert_success_message(
-                'Cloud Tenant "{}" created'.format(self.name))
+                'Cloud Tenant "{}" created'.format(name))
+
+        tenant = Tenant(name, provider)
```
Comment (mfalesni) [inline cfme/cloud/tenant.py]: This should use the `instantiate` here.

Comment (vatsalparekh) [inline cfme/cloud/tenant.py]: Fixed.

**Comment by dajoRH** [comment]
Lint report for commit 7ec9ffda5ee13b1869ac8e3ee29e8114d0144c1e:

`cfme/cloud/tenant.py`:
- :red_circle: Line 7:1: **F401** *'navmazing.NavigateToObject' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/cloud/tenant.py [hunk 21]
```diff
@@ -12,7 +12,6 @@
 from widgetastic_manageiq import (
     Accordion, BootstrapSelect, BreadCrumb, ItemsToolBarViewSelector, PaginationPane, Search,
     SummaryTable, Table, Text)
-
```
Comment (psav) [inline cfme/cloud/tenant.py]: **REQUIRED**: Please leave this blank line

File: cfme/cloud/tenant.py [hunk 22]
```diff
@@ -22,7 +21,6 @@
 from utils.wait import wait_for, TimedOutError
 from utils.version import Version
 
-
```
Comment (psav) [inline cfme/cloud/tenant.py]: **REQUIRED**: And this one

File: cfme/tests/cloud/test_tenant.py [hunk 23]
```diff
@@ -13,7 +13,8 @@
 
 @pytest.yield_fixture(scope='function')
 def tenant(provider, setup_provider):
-    tenant = Tenant(name=fauxfactory.gen_alphanumeric(8), provider=provider)
+    collection = TenantCollection()
```
Comment (psav) [inline cfme/tests/cloud/test_tenant.py]: **REQUIRED**: We should always pass an appliance object to a collection, this is fundamental to the migration to collections. The fixture can just add appliance on line 15 and it will get it via a global fixture.

File: cfme/cloud/tenant.py [hunk 24]
```diff
@@ -194,48 +192,97 @@ def is_displayed(self):
             self.entities.breadcrumb.active_location == 'Tag Assignment')
 
 
-class Tenant(Navigatable):
-    _param_name = 'Tenant'
+class TenantCollection(Navigatable):
+    """Collection object for the :py:class:`cfme.cloud.tenant.Tenant`."""
```
Comment (psav) [inline cfme/cloud/tenant.py]: **REQUIRED**: As per below, We need to have an ```__init__``` which takes appliance and saves it to ```self.appliance```
